### ISSUE TITLE
Drupal Behaviors being attached twice

### ISSUE THREAD

**Issue by q0rban** [e0ipso/js_defer#13]
I'm noticing that when using JS Defer module, our behaviors are getting attached twice. This happens whether JS aggregation is on or off.

**Comment by q0rban** [comment]
Ok, it appears that any JS file that defines a Drupal behavior and that is _not_ being deferred gets its behaviors attached twice. This can lead to some oddities, as it is possible for event handlers to get attached to elements more than once.

**Comment by wesruv** [comment]
I missed the mark on some of the implementation. Looking into this.

**Comment by e0ipso** [comment]
Thank you guys!

**Comment by wesruv** [comment]
This is the line that's doing it, it's intentionally sending document as a context again:
https://github.com/e0ipso/js_defer/blob/7.x-1.x/js/lazy-load-js.js#L42

Makes sense because the lazy loaded elements won't have had their behaviors run.

I'm going to try something where we see what behaviors exist before we lazy load, and only load the ones that didn't exist after. I think that's a good idea?

**Comment by q0rban** [comment]
> I'm going to try something where we see what behaviors exist before we lazy load, and only load the ones that didn't exist after. I think that's a good idea?

Makes sense to me!

**Comment by wesruv** [comment]
For our needs I'm overriding Drupal.js_defer_load to do this:

``` js
      var existingBehaviors = Object.keys(Drupal.behaviors);

      LazyLoad.js(Drupal.settings.js_defer[queueName].scripts, function () {
        // Execute only new behaviors since we ran the queue
        $.each(Drupal.behaviors, function (index) {
          if ($.inArray(index, existingBehaviors) < 0 && $.isFunction(this.attach)) {
            this.attach(document, Drupal.settings);
          }
        });
      });
```

As we talked about today though, I'm not sure that this is something your module needs to do, as we should be wrapping things that should only get run once per element with `.once()`.

If you'd like I can make a PR, otherwise we'll just go with our override :smile:

**Comment by q0rban** [comment]
I think this is important to do in the js_defer module. There are many behaviors that intentionally don't use `once()` because it would break the expected functionality of the behavior.

**Comment by e0ipso** [comment]
After thinking about this I think to remember that this is by design.

The deferred script may change the DOM (for instance) to add an `<article>`. If an already existing behavior modifies all `<article>` it needs to be re-run to have the chance to interact with the deferred content.

This is something that happens very often in Drupal core:
- Behaviors are attached every time an Ajax call is made (using the PHP ajaxy helpers)
- Behaviors are attached every time a pager link is clicked in ajax Views.
- …

In general, if your behavior is not safe to run multiple times then it's your behavior's responsibility to safe guard against that, because Drupal can run them as many times as needed so everyone can act upon the latest ajax/deferred changes.

Does that make sense?

**Comment by q0rban** [comment]
If a behavior modifies the DOM without using the traditional means (Drupal.ajax), then it is the responsibility of that script to re-attach behaviors to _just_ the portion of DOM that is modified. (For an example of a script doing this, see [the JS for Views Load More](http://cgit.drupalcode.org/views_load_more/tree/views_load_more.js?h=7.x-1.x#n99)). The way JS Defer module does it is running all behaviors twice on the same context, i.e. `document`. This is not the expected behavior for Behaviors, IMO.

**Comment by e0ipso** [comment]
I find [this comment](http://cgit.drupalcode.org/drupal/tree/misc/drupal.js?h=7.x#n29) specially relevant. 

```
 * Behaviors should use
 * @code
 *   $(selector).once('behavior-name', function () {
 *     ...
 *   });
 * @endcode
 * to ensure the behavior is attached only once to a given element.
```

Seems to indicate that the responsibility is in the implementor.

```
 *   An element to attach behaviors to. If none is given, the document element
 *   is used.
```

Suggests that the document is a valid context.

---

> re-attach behaviors to just the portion of DOM that is modified

Since there is no way the `js_defer` module can know what the deferred JS file is doing, I can't see how to narrow down the context for the behavior re-attaching. Do you have a possible implementation for this in mind?

**Comment by q0rban** [comment]
>  Do you have a possible implementation for this in mind?

Yes. Performance. :)

**Comment by q0rban** [comment]
I should say, I do agree that a module author _should_ do their best to ensure that their JS only runs once. I'm not sure we can safely assume that all of Contrib ascribes to this ideology. Indeed, even a quick perusal through a few JS files in Views and CTools shows some examples where functionality would run twice with JS Defer module on:
- http://cgit.drupalcode.org/views/tree/js/ajax.js#n134
- http://cgit.drupalcode.org/views/tree/js/ajax_view.js#n14
- http://cgit.drupalcode.org/views/tree/js/base.js#n30
- http://cgit.drupalcode.org/ctools/tree/js/dependent.js#n204
- http://cgit.drupalcode.org/ctools/tree/js/stylizer.js#n204

But to elaborate on the performance implications, setting instances like the above aside…even if all of Contrib used `$.once`, think about the fact that `$.once` is called _after_ jQuery has parsed through the DOM to find the elements to check. And that this likely happens in every Behavior, perhaps multiple times, and that if the Behavior is modifying multiple elements on the page, jQuery would need to iterate through each one, checking the `once` status on them all over again.

**Comment by wesruv** [comment]
FWIW, I'm torn on this as a patch to the module vs something we only do on our project.

I think implementors should be using `.once()` on anything that can break if it's run twice on the same element, and I think the module calling all of the `Drupal.behaviors` with a `document` context is valid.

On the other hand, for our specific use it's being run on every page because Drupal wasn't deferring JS like it should after it's been aggregated. I don't think this was necessarily how this module was intended to be used, but I don't feel like I'm misusing it.

In our use case I think we're wise to make sure that `Drupal.behaviors` isn't run twice on the document because:
- The DOM is kind of a wreck already, I don't want to make the browser work any harder than I have to
- I have a very good idea what's in the deferred scripts, and I don't think it's anything that needs to have behaviors run on the document after they run
- I don't trust all of the custom JS on the site and I don't have the time to refactor all of it

James makes a good point about the standard vs the reality, although I can't speak to the instances he points out. Would the second run with a `document` context actually cause bugs, just awkwardness, or no issues at all. From the look of them, I'm thinking awkwardness at a minimum.

We're able to replace the relevant code in the Drupal.behaviors object, so it's not like we had to hack the module to get this to work the way we wanted.

**Comment by e0ipso** [comment]
> Do you have a possible implementation for this in mind?
> 
> Yes. Performance. :)

I meant: how would you to narrow down the scopes for the deferred files? So we don't have to use `document`.

**Comment by q0rban** [comment]
@e0ipso and I chatted about this today and he brought up a good point. Some of these deferred JS files may not be Drupal behaviors. They may be adding markup to the page that other behaviors need to respond to. For example, if you deferred the JS file that brings Facebook comments onto the page, but then there are Drupal behaviors that need to interact with the new markup on the page, the JS Defer module is doing the correct thing by re-attaching all behaviors.

For this reason, we decided it would be best to add an additional configuration option to each queue in `hook_js_defer_info()`. The option could be `reattach_all_behaviors`, and would default to `true`. Then, in `Drupal.js_defer_load`, we could check the value of that option and use @wesruv's method of only attaching new Behaviors if it is `false`.

**Comment by wesruv** [comment]
I'm on it

**Comment by wesruv** [comment]
PR: https://github.com/e0ipso/js_defer/pull/15

Tested this on our current project. If you upgrade the module from current state, nothing should change. This is only adding an option.

@e0ipso

**Comment by e0ipso** [comment]
Tagged and bagged: https://www.drupal.org/project/js_defer/releases/7.x-1.1

Thanks all!
